### PR TITLE
Support for custom mac address

### PR DESCRIPTION
**Pull request by jfparis**
This commit adds support for custom mac addresses previously implanted with "espefuse.py burn_custom_mac"

## What does this implement/fix?

ESP32 allows user to burn a custom MAC address via espefuse.py
https://docs.espressif.com/projects/esptool/en/latest/esp32/espefuse/burn-custom-mac-cmd.html
This is especially useful when a batch of chips comes out with identical MAC or a blank (00:..:00) MAC
This PR support such MAC

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/core/helpers.cpp [hunk 1]
```diff
@@ -443,7 +443,8 @@ void get_mac_address_raw(uint8_t *mac) {  // NOLINT(readability-non-const-parame
   // without doing the CRC check.
   esp_efuse_read_field_blob(ESP_EFUSE_MAC_FACTORY, mac, 48);
 #else
-  esp_efuse_mac_get_default(mac);
+  if (esp_efuse_mac_get_custom(mac) == ESP_OK) {}
+  else esp_efuse_mac_get_default(mac);
```
Comment (ssieb) [inline esphome/core/helpers.cpp]:
```suggestion
  if (esp_efuse_mac_get_custom(mac) != ESP_OK)
    esp_efuse_mac_get_default(mac);
```

Comment (jfparis) [reply esphome/core/helpers.cpp]: Thank you. Fixed that

**Comment by jfparis** [comment]
Hello. Is there anything I can do / change to help this into the main branch?

File: esphome/core/helpers.cpp [hunk 2]
```diff
@@ -443,7 +443,8 @@ void get_mac_address_raw(uint8_t *mac) {  // NOLINT(readability-non-const-parame
   // without doing the CRC check.
   esp_efuse_read_field_blob(ESP_EFUSE_MAC_FACTORY, mac, 48);
 #else
-  esp_efuse_mac_get_default(mac);
+  if (esp_efuse_mac_get_custom(mac) != ESP_OK)
```
Comment (mmakaay) [inline esphome/core/helpers.cpp]: Wouldn't it be more useful to  implement this as follows?

```c++
void get_mac_address_raw(uint8_t *mac) {  // NOLINT(readability-non-const-parameter)
#if defined(USE_ESP32)
  if (esp_efuse_mac_get_custom(mac) != ESP_OK) {
#if defined(USE_ESP32_IGNORE_EFUSE_MAC_CRC)
    // On some devices, the MAC address that is burnt into EFuse does not
    // match the CRC that goes along with it. For those devices, this
    // work-around reads and uses the MAC address as-is from EFuse,
    // without doing the CRC check.
    esp_efuse_read_field_blob(ESP_EFUSE_MAC_FACTORY, mac, 48);
#else
    esp_efuse_mac_get_default(mac);
#endif
  }
#elif defined(USE_ESP8266)
```

That would bring custom MAC support to all devices, not only those for which `USE_ESP32_IGNORE_EFUSE_MAC_CRC` is not defined.

Comment (jfparis) [reply esphome/core/helpers.cpp]: This makes sense. Thank you for reviewing

**Pull request closed without merging**